### DISCUSSION TITLE
Modernizing code

### DISCUSSION THREAD

**Discussion by nwalters512** [PrairieLearn/PrairieLearn#8874] [Dev]
We've been working on modernizing the PrairieLearn codebase for a while now, but it's not always clear what "modernization" entails. This note tries to summarize exactly what modernization means, and provides a suggested sequence for modernizing code.

## What does modern code look like?

In no particular order:

### Replacing callbacks with `async`/`await`

"callback hell" is a very real problem. Libraries like [`async`](https://github.com/caolan/async) can make it more easy to deal with, but nothing beats native async code for its conciseness, reliable error handling, and ergonomics.

```ts
// In Express route handlers:
import asyncHandler = require('express-async-handler')
router.get(asyncHandler(async (req, res) => {
  const data = await doThing();
  res.json(data);
}));
```

### TypeScript

Types are great; they help catch mistakes as we make them and allow us to more easily and confidently refactor code. Type checking can be enabled by adding `// @ts-check` to the top of `*.js` files, or (preferably) by renaming `*.js` files to `*.ts`.

### Parsed/validated data

This goes hand-in-hand with TypeScript: we want to make sure that data entering our system from outside sources has the expected properties, shape, types, etc. Our library of choice for this is [`zod`](https://zod.dev/).

The most common use case for this is SQL queries. We have a number of functions that will automatically parse query results with a Zod schema and ensure that the results are property typed:

```ts
import { queryRows } from '@prairielearn/postgres';
import { z } from 'zod';

const UserSchema = z.object({
  name: z.string(),
  age: z.number(),
});

const result = await queryRows('SELECT * FROM users;', {}, UserSchema);
```

### Replacing CJS with ESM syntax

CommonJS has served us well, but we're starting to prepare for an ESM-first future. To that end, we're replacing CJS with ESM syntax where we can. At the moment, we're still transpiling ESM back to CJS at runtime, but converting everything to at least be authored in ESM will eventually make the transition to native ESM at runtime relatively straightforward. This transpilation step means there are still some rough edges, so constructs like `import foo = require('foo')` are sometimes required.

### Generating HTML in TypeScript

We've historically used [EJS](https://github.com/mde/ejs) as our templating language for responses. However, it's impossible to use type checking with EJS, so we've started using our own homegrown [`@prairielearn/html`](https://github.com/PrairieLearn/PrairieLearn/tree/master/packages/html) library to generate HTML directly in TypeScript.

### Avoiding the use of `res.locals`

`res.locals` is difficult to type check and use. Where possible, prefer passing around well-typed objects instead of `res.locals` as a whole. And in route handlers, prefer querying directly instead of attaching data to `res.locals` in a middleware and reading it in the route handler.

We're not particularly dogmatic about this. While we want to avoid `res.locals` in brand-new code, we acknowledge that it will be a long journey to remove it from existing code. So, don't feel like you have to jump through hoops to avoid it, and if you have to use it in new code to work with existing code, that's fine too.

## Recommended sequence

1. Identify one or more files you wish to modernize.
2. If they don't already have type checking enabled, open and merge a PR adding `// @ts-check` and making the minimum number of changes to get the TypeScript compiler happy with them.
3. Open and merge a PR converting the files from CJS to ESM.
4. Open and merge a PR converting to Zod-validated queries and replacing callback-based functions with `async`/`await` ones.
5. Open and merge a PR converting the file to TypeScript (rename to have `.ts` file extension and replace JSDoc types with native types).
6. Open and merge a PR converting EJS templates to use `html` tagged template literals. You may also do this as a part of the above step if you wish.

**Comment by jonatanschroeder** [comment]
I'd recommend that item 4 above be split into multiple parts, since async/await changes and validated queries tend to cause a lot of file changes (in particular with the reduction of indentation requirements) that end up losing history due to renaming. This is not as big a deal in pages, but it's a bigger concern in libs. So:

4. Open and merge a PR converting queries to Zod-validated queries and replacing callback-based functions with async/await ones.
5. Open and merge a PR converting the file to TypeScript (.ts file extension, replacing JSDoc types with native types).
6. ... EJS to html

**Reply by nwalters512** [reply]
Great call - I had already forgotten about the problems posed by doing the rename and large changes in the same PR! I'll update the sequence above.

**Comment by mwest1066** [comment]
A few more things we could add:

### Use the updated query functions

The [prairielearn/postgres](https://github.com/PrairieLearn/PrairieLearn/tree/master/packages/postgres) package has been updated with new functions for making validated queries:
* `const row = queryRow(sql, params, schema)` -- get a single row that matches `schema`
* `const rowOrNull = queryOptionalRow(sql, params, schema)` -- get one row matching `schema`, or `null`
* `cost rows = queryRows(sql, params, schema)` -- get zero or more rows as an array, each of which match `schema`

These new functions should be used in preference to the `queryValidated*()` functions, which should no longer be used in new code. When modernizing it is clean swap to replace these functions as follows:
* `queryRow` is a drop-in replacement for `queryValidatedOneRow` and `queryValidatedSingleColumnOneRow`
* `queryOptionalRow` is a drop-in replacement for `queryValidatedZeroOrOneRow` and `queryValidatedSingleColumnZeroOrOneRow`
* `queryRows` is a drop-in replacement for `queryValidatedRows` and `queryValidatedSingleColumnRows`

Note that the `callValidated*()` functions should still be used when calling stored procedures, but they should eventually be eliminated as we transition away from stored procedures.

### Move code out of SQL sprocs and into TypeScript

PrairieLearn historically made heavy use of PostgreSQL stored procedures ("sprocs"). These allowed up to use a strongly-typed language for some of the more critical pieces of code (e.g., authentication, grading, syncing) when the rest of the code base was in JavaScript.

Now that we are converting to TypeScript we want to rewrite our sprocs in TypeScript as well. This makes the code easier to read, test, and debug. As a general rule, a sproc should be converted to a single TypeScript function where all the logic is shifted into TypeScript and only the basic SQL queries are kept in SQL. To preserve the transactional nature of sprocs it is common to wrap the contents of the replacement TypeScript function in a `runInTransactionAsync()` call.

Our eventual goal is to have zero sprocs in PL. This will also help us to manage DB connections using connection pooling between servers, which is problematic when using sprocs.

### Use "DB types" whenever possible

The file [`lib/db-types.ts`](https://github.com/PrairieLearnInc/PrairieTest/blob/main/src/lib/db-types.ts) contains canonical `zod` schema and the corresponding TypeScript types for all SQL tables. Whenever possible we want to return data from DB queries using these schema and we want to write TypeScript functions that take arguments using these types (or combinations of them).

We use a mixed casing convention where the DB types are in CamelCase but objects of these basic types are in `snake_case` to match the SQL convention. For example, a row from the DB table `grading_jobs` will have `zod` schema `GradingJobSchema`, TypeScript type `GradingJob`, and object name `grading_job`. We use plural snake case like `grading_jobs` for an array of such objects.

In general we don't make derived types of the DB types to store extra data. Instead make a new container type (and schema) to store the extra data. For example, if we have a query that returns a list of courses and their enrollment counts then we would make a type like:
```
interface CourseWithEnrollment {
  course Course;
  enrollment_count number;
}
```

### Move helper code into `models/`

We are separating out the code in `lib/` into helper code that deals with a single table, which should go into the new `models/` directory, and more complicated multi-table code that should stay in `lib/`. The code in `models/` should always be in a file named using kebab-case after the corresponding table name. For example, code dealing with `grading_jobs` would is in `models/grading-jobs.ts` (and possibly `models/grading-jobs.sql`).

As a (currently) hypothetical example, the grading code manipulates data in many tables, including `questions`, `instance_questions`, `assessment_instances`, etc. Code that deals with just one of these tables (e.g., a `selectQuestionById()` function) would go into `models/questions.ts`, while the overall grading code would go into `lib/grading.ts` (and `lib/grading.sql`).

### Format dates and intervals in TypeScript

To ensure consistent date/time/interval formatting we originally used the PostgreSQL formatting functions in SQL. With the [Intl](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl) API now available in JavaScript we want to move all date/time/interval formatting into JavaScript. All date, time, and interval data should be exported from SQL as raw types and formatted using the `formateDate()`, `formatInterval()`, etc., functions.

### Move tests next to code

Testing code has been historically kept in the `apps/prairielearn/src/tests` directory. Instead we are moving it next to the code that it tests, using a `.test.ts` suffix. For example, the tests for `models/grading-jobs.ts` would now go in `models/grading-jobs.test.ts` (and possibly in `models/grading-jobs.test.sql` if needed). Keeping testing code close to the code being tested makes it easier to bounce back and forwards between the two.

**Comment by mwest1066** [comment]
## Additions to the recommended sequence

After step 6 (convert EJS to `html` tagged templates) the next steps are:

7. Rewrite SQL queries to return whole-row types from `db-types.ts`, rather than custom types per file. We want to eliminate complex custom types and instead have custom types which are just collections of types from `db-types.ts`.
8. Use the updated query functions (`queryRows()`, etc).
9. Format dates in TypeScript.
10. Replace `sprocs` with TypeScript code wherever possible, writing new TypeScript helper functions as needed.
